upgrades/steps_28.go

```diff
@@ -139,59 +156,163 @@ var getUpgradeStepsClient = func(caller base.APICaller) UpgradeStepsClient {
 	}
 
 	client := getUpgradeStepsClient(ctx.APIState())
 	if err := client.WriteUniterState(uniterStates); err != nil {
 		return nil, errors.Annotatef(err, "unable to set state for units %q", strings.Join(unitNames, ", "))
 	}
-	return fileNames, nil
+	return fileNames.Values(), nil
+}
+
+func readUniterState(uniterStateDir string) (*string, string, error) {
+	uniterStateFile := filepath.Join(uniterStateDir, "uniter")
+	// Read into a State as we can validate the data.
+	var st operation.State
+	if err := utils.ReadYaml(uniterStateFile, &st); err != nil {
+		// No file available, move on.
+		if os.IsNotExist(err) {
+			return nil, "", err
+		}
+		return nil, "", errors.Annotatef(err, "failed to read uniter state from %q", uniterStateFile)
+	}
+	if err := st.Validate(); err != nil {
+		return nil, "", errors.Annotatef(err, "failed to validate uniter state from %q", uniterStateFile)
+	}
+
+	data, err := yaml.Marshal(st)
+	if err != nil {
+		return nil, "", errors.Annotatef(err, "failed to unmarshal uniter state from %q", uniterStateFile)
+	}
+	str := string(data)
+
+	return &str, uniterStateFile, nil
 }
 
 func caasOperatorLocalState(ctx Context, unitNames []string) ([]string, error) {
 	// TODO: (hml) 27-Mar-2020.
 	// Implement when relations, storage, metrics and the operator state are moved.
 	return nil, nil
 }
+
+func readStorageState(uniterStateDir string) (*string, string, error) {
+	storageStateDir := filepath.Join(uniterStateDir, "storage")
+	storage, err := readAllStateFiles(storageStateDir)
+	if err != nil {
+		return nil, "", errors.Annotatef(err, "cannot load storage state from %q", storageStateDir)
+	}
+	if len(storage) < 1 {
+		return nil, "", nil
+	}
+
+	data, err := yaml.Marshal(storage)
+	if err != nil {
+		return nil, "", errors.Annotatef(err, "failed to unmarshal storage state from %q", storageStateDir)
+	}
+	str := string(data)
+
+	return &str, storageStateDir, nil
+}
+
+// readAllStateFiles loads and returns every stateFile persisted inside
+func readAllStateFiles(dirPath string) (map[string]bool, error) {
+	if _, err := os.Stat(dirPath); os.IsNotExist(err) {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	fis, err := ioutil.ReadDir(dirPath)
+	if err != nil {
+		return nil, err
+	}
+	files := make(map[string]bool)
+	for _, fi := range fis {
+		if fi.IsDir() {
+			continue
+		}
+		storageId := fi.Name()
+		if i := strings.LastIndex(storageId, "-"); i > 0 {
+			storageId = storageId[:i] + "/" + storageId[i+1:]
+			if !names.IsValidStorage(storageId) {
+				continue
+			}
+		} else {
+			// Lack of "-" means it's not a valid storage ID.
+			continue
+		}
+		tag := names.NewStorageTag(storageId)
+		f, err := readStateFile(dirPath, tag)
+		if err != nil {
+			return nil, err
+		}
+		files[tag.Id()] = *f
+	}
+	return files, nil
+}
+
+// readStateFile loads a stateFile from the subdirectory of dirPath named
+// is returned.
+func readStateFile(dirPath string, tag names.StorageTag) (*bool, error) {
```

Comment:
Same comment about returning pointers to scalars.

Reply:
done